File: backend/app/routers/endpoints.py
"""Endpoint Security: host inventory, telemetry derived from logs, containment."""
from __future__ import annotations

import sqlite3

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from ..db import get_db, row, rows

router = APIRouter(prefix="/endpoints", tags=["endpoints"])
SHELLS = ("%powershell.exe", "%cmd.exe", "/bin/bash", "/bin/sh", "/usr/bin/sudo")


class ContainBody(BaseModel):
    contained: bool


def _get(db, ep_id: int) -> dict:
    e = row(db.execute("SELECT * FROM endpoints WHERE id=?", (ep_id,)).fetchone(), ("browser_history",))
    if not e:
        raise HTTPException(404, "Endpoint not found")
    e["contained"] = bool(e["contained"])
    return e


@router.get("")
def list_endpoints(search: str | None = None, db: sqlite3.Connection = Depends(get_db)):
    where, params = "1=1", []
    if search:
        where = "(hostname LIKE ? OR ip_address LIKE ? OR primary_user LIKE ?)"
        params = [f"%{search}%"] * 3
    out = rows(db.execute(
        "SELECT e.id, e.hostname, e.ip_address, e.os, e.primary_user, e.domain, e.last_seen, e.contained, "
        "(SELECT COUNT(*) FROM alerts a WHERE a.details LIKE '%' || e.hostname || '%' AND a.status != 'closed') "
        f"AS open_alerts FROM endpoints e WHERE {where} ORDER BY e.hostname", params))
    for e in out:
        e["contained"] = bool(e["contained"])
    return out


@router.get("/{ep_id}")
def get_endpoint(ep_id: int, db: sqlite3.Connection = Depends(get_db)):
    e = _get(db, ep_id)
    h = e["hostname"]
    e["processes"] = rows(db.execute(
        "SELECT id, timestamp, process, command_line, username, raw_log FROM logs "
        "WHERE hostname=? AND type='OS' AND process IS NOT NULL ORDER BY timestamp DESC LIMIT 100", (h,)))
    e["network"] = rows(db.execute(
        "SELECT id, timestamp, type, source_address, source_port, destination_address, destination_port, "
        "process, action FROM logs WHERE hostname=? AND type IN ('Network','Proxy','DNS') "
        "ORDER BY timestamp DESC LIMIT 100", (h,)))
    shell_clause = " OR ".join("process LIKE ?" for _ in SHELLS)
    e["terminal_history"] = rows(db.execute(
        f"SELECT id, timestamp, process, command_line, username FROM logs WHERE hostname=? AND type='OS' "
        f"AND command_line IS NOT NULL AND ({shell_clause}) ORDER BY timestamp DESC LIMIT 50", (h, *SHELLS)))
    e["alerts"] = rows(db.execute(
        "SELECT id, event_id, rule_name, severity, status, created_at FROM alerts "
        "WHERE details LIKE ? ORDER BY created_at DESC", (f"%{h}%",)))
    return e


@router.post("/{ep_id}/containment")
def set_containment(ep_id: int, body: ContainBody, db: sqlite3.Connection = Depends(get_db)):
    _get(db, ep_id)
    db.execute("UPDATE endpoints SET contained=? WHERE id=?", (int(body.contained), ep_id))
    db.commit()
    return _get(db, ep_id)

File: frontend/js/views/endpoints.js
import { api } from "../api.js";
import { debounce, emptyState, html, icon, isoTz, loading, logsLink, modal, setHTML, sevBadge, toast } from "../utils.js";

const TABS = [
  ["processes", "Processes"], ["network", "Network Action"], ["terminal", "Terminal History"],
  ["browser", "Browser History"], ["alerts", "Alerts"],
];

export function mount(el, params) {
  const state = { items: [], selected: null, host: params.host || null, search: "", detail: null, tab: "processes" };

  setHTML(el, html`
    <div class="page-head"><div><h1 class="page-title">Endpoint Security</h1>
      <p class="page-sub">EDR view of managed hosts: process tree, network activity, command history and containment.</p></div></div>
    <div class="split">
      <section class="panel">
        <div class="list-tools"><div class="search-input">${icon("search")}<input class="input" data-search placeholder="Hostname, IP or user…"></div></div>
        <div class="list" id="epList">${loading()}</div>
      </section>
      <section class="panel detail" id="epDetail"></section>
    </div>`);
  const listEl = el.querySelector("#epList"), detailEl = el.querySelector("#epDetail");

  async function loadList() {
    try { state.items = await api.get("/endpoints", { search: state.search }); }
    catch (e) { setHTML(listEl, html`<div class="error-box">${e.message}</div>`); return; }
    if (state.host) {
      state.selected = state.items.find((i) => i.hostname === state.host)?.id ?? null;
      state.host = null;
    }
    if (!state.selected && state.items.length) state.selected = state.items[0].id;
    renderList(); loadDetail();
  }

  function renderList() {
    setHTML(listEl, state.items.length ? state.items.map((e) => html`
      <div class="list-item ${e.id === state.selected ? "sel" : ""}" data-id="${e.id}" tabindex="0">
        <div class="row1"><span class="title">${e.hostname}</span>
          ${e.contained ? html`<span class="badge red">${icon("lock")} Contained</span>` : html`<span class="badge green">Online</span>`}</div>
        <div class="meta"><span class="mono">${e.ip_address}</span><span>${e.os}</span>
          ${e.open_alerts ? html`<span class="sev sev-High">${e.open_alerts} open alert${e.open_alerts > 1 ? "s" : ""}</span>` : ""}</div>
      </div>`) : emptyState("No endpoints found", "hdd"));
  }

  async function loadDetail() {
    if (!state.selected) { setHTML(detailEl, emptyState("Select an endpoint", "hdd")); return; }
    setHTML(detailEl, loading());
    try { state.detail = await api.get(`/endpoints/${state.selected}`); }
    catch (e) { setHTML(detailEl, html`<div class="error-box">${e.message}</div>`); return; }
    renderDetail();
  }

  function tabBody(d) {
    if (state.tab === "processes") {
      return d.processes.length ? html`<div class="table-wrap"><table class="table compact lined"><thead><tr><th>Time</th><th>Process</th><th>Command line</th></tr></thead>
        <tbody>${d.processes.map((p) => html`<tr class="row"><td class="mono" style="white-space:nowrap">${p.timestamp}</td><td class="mono">${p.process}</td><td class="mono">${p.command_line || ""}</td></tr>`)}</tbody></table></div>`
        : emptyState("No process telemetry");
    }
    if (state.tab === "network") {
      return d.network.length ? html`<div class="table-wrap"><table class="table compact lined"><thead><tr><th>Time</th><th>Type</th><th>Destination</th><th>Process</th><th>Action</th></tr></thead>
        <tbody>${d.network.map((n) => html`<tr class="row"><td class="mono" style="white-space:nowrap">${n.timestamp}</td><td>${n.type}</td>
          <td class="mono"><a href="${logsLink(`destination_address="${n.destination_address}"`)}">${n.destination_address}</a>${n.destination_port ? ":" + n.destination_port : ""}</td>
          <td class="mono">${n.process || "-"}</td><td>${n.action}</td></tr>`)}</tbody></table></div>`
        : emptyState("No network telemetry");
    }
    if (state.tab === "terminal") {
      return d.terminal_history.length ? html`<div class="raw-block">${d.terminal_history.map((t) => `[${t.timestamp}] ${t.username || ""}> ${t.command_line}\n`)}</div>`
        : emptyState("No shell commands recorded", "terminal");
    }
    if (state.tab === "browser") {
      return d.browser_history.length ? html`<table class="table compact lined"><thead><tr><th>Time</th><th>URL</th></tr></thead>
        <tbody>${d.browser_history.map((b) => html`<tr class="row"><td class="mono">${b.time}</td><td class="mono">${b.url}</td></tr>`)}</tbody></table>`
        : emptyState("No browser history");
    }
    return d.alerts.length ? html`<table class="table compact lined"><thead><tr><th>Severity</th><th>Rule</th><th>Status</th><th>Date</th></tr></thead>
      <tbody>${d.alerts.map((a) => html`<tr class="row"><td>${sevBadge(a.severity)}</td><td>${a.rule_name}</td><td>${a.status}</td><td class="mono">${isoTz(a.created_at)}</td></tr>`)}</tbody></table>`
      : emptyState("No alerts for this host", "checkCircle");
  }

  function renderDetail() {
    const d = state.detail;
    setHTML(detailEl, html`
      <div class="detail-head"><div><h2>${d.hostname}</h2>
        <div class="muted small">${d.os} · Last seen ${d.last_seen ? isoTz(d.last_seen) : "-"}</div></div>
        <div class="btn-row">
          <a class="btn btn-sm" href="${logsLink(`hostname="${d.hostname}"`)}">${icon("list")} Logs</a>
          <button class="btn btn-sm ${d.contained ? "" : "btn-danger"}" data-contain>${icon(d.contained ? "unlock" : "lock")} ${d.contained ? "Release Containment" : "Contain Host"}</button>
        </div></div>
      <div class="kv">
        <div class="k">IP Address</div><div class="v mono">${d.ip_address}</div>
        <div class="k">Primary User</div><div class="v">${d.domain && d.domain !== "-" ? `${d.domain}\\${d.primary_user}` : d.primary_user}</div>
        <div class="k">Status</div><div class="v">${d.contained ? html`<span class="badge red">Contained - network isolated</span>` : html`<span class="badge green">Online</span>`}</div>
      </div>
      <div class="subtabs" role="tablist">${TABS.map(([k, l]) => html`<button role="tab" class="${state.tab === k ? "active" : ""}" data-tab="${k}">${l}</button>`)}</div>
      <div style="padding-top:12px">${tabBody(d)}</div>`);
  }

  el.addEventListener("click", (e) => {
    const item = e.target.closest(".list-item");
    if (item) { state.selected = Number(item.dataset.id); renderList(); loadDetail(); return; }
    const tab = e.target.closest("[data-tab]");
    if (tab) { state.tab = tab.dataset.tab; renderDetail(); return; }
    if (e.target.closest("[data-contain]")) {
      const d = state.detail, contain = !d.contained;
      modal({
        title: contain ? "Contain host?" : "Release containment?",
        body: html`<p>${contain ? "Network-isolate" : "Restore network access for"} <b>${d.hostname}</b> (${d.ip_address}).</p>`,
        confirmText: contain ? "Contain" : "Release", danger: contain,
        onConfirm: async () => {
          state.detail = { ...state.detail, ...(await api.post(`/endpoints/${d.id}/containment`, { contained: contain })) };
          toast(`${d.hostname} ${contain ? "contained" : "released"}`);
          renderDetail();
          state.items = await api.get("/endpoints", { search: state.search }); renderList();
        },
      });
    }
  });
  el.addEventListener("keydown", (e) => { if (e.key === "Enter" && e.target.matches(".list-item")) e.target.click(); });
  const onSearch = debounce(() => { state.selected = null; loadList(); }, 300);
  el.addEventListener("input", (e) => { if (e.target.matches("[data-search]")) { state.search = e.target.value; onSearch(); } });

  loadList();
}
